### PR TITLE
Added files required to schedule mails for sending invoices to clients

### PR DESCRIPTION
**Pull request by chadivesatish**
Requirement : To send invoice mails to clients as per the schedules taken from user.
Tasks finished till now : 
1.Initiate cronjobs at 6AM everyday and fetch all the schedules that are expected to work on today.
2.Fetch all the invoices to send as per the invoice data recorded in those schedules.
3.Print all the invoices data (in pdfs ).
4.Send these printed pdfs through mail to the client.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -88,6 +88,6 @@ app.use(function (err, req, res, next) {
     res.render("error");
 });
 
-
+var msg = require("./scheduleCron.js");
```
Comment (faisal3389) [inline app.js:91]: are we using this msg in this file, if not why import it?

File: scheduleCron.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
+/* eslint-disable no-unused-vars */
+/* eslint-disable no-undef */
```
Comment (faisal3389) [inline scheduleCron.js:2]: why do we need these exceptions?

File: scheduleCron.js [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+var express = require("express");
+var fs = require("fs");
+var handlebars = require("handlebars");
+const puppeteer = require("puppeteer");
+const ScheduleStatus = require("./models/scheduleStatus");
+var scheduleStatus=[] ;
+
+ 
+function validateStartDate(date){
+    const currDate = new Date();
```
Comment (faisal3389) [inline scheduleCron.js:18]: use moment module, it will do all this in one line
`date.isBefore(new Date())`

File: scheduleCron.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        }
+    }
+    else{
+        return false;
+    }
+}
+const filterSchedules = async () =>{
+    var filteredSchedules=[];
+    var scheduleFilter = await schedule.find({isActive: true});
+    scheduleFilter.filter(function(item) {
```
Comment (faisal3389) [inline scheduleCron.js:46]: use ES6 functions

File: models/scheduleStatus.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    pdfPrintStatus: {
+        type: String,
+        required: [true, "please enter pdfPrintStatus"]
+    },
+
+    sendMailStatus: {
+        type: String,
+        required: [true, "please Enter sendMailStatus"]
+    },
```
Comment (faisal3389) [inline models/scheduleStatus.js:36]: lets use only one status with success and fail for each step

File: scheduleCron.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            filteredSchedules.push(item);
+        }
+    });
+    scheduleFilter.filter(function(item) {
+        if(item.frequency=== "Anually" && item.isDisabled === false && validateStartDate(item.date) && new Date(item.date).getDate() === new Date().getDate() && new Date(item.date).getMonth() === new Date().getMonth() ){
+            scheduleStatus.push({date: new Date().toDateString() ,scheduleId: item._id, invoiceId: "",invoiceFetchStatus: "false",pdfPrintStatus: "false", sendMailStatus: "false"});
+            filteredSchedules.push(item);
+        }
+    });
+    return filteredSchedules;
```
Comment (faisal3389) [inline scheduleCron.js:70]: `
 scheduleFilter
    .filter(item.isDisabled && validateStartDate(item.date))
    .filter((item) => {
        if (isValidSchedule(item)) {
            scheduleStatus.push({date: new Date().toDateString() ,scheduleId: item._id, invoiceId: "",invoiceFetchStatus: "false",pdfPrintStatus: "false", sendMailStatus: "false"});
            filteredSchedules.push(item);
        }
     });



     isValidSchedule(item) => {
        if(item.frequency=== "Daily") {
            return;
        } else if (item.frequency=== "Monthly") {
            return new Date(item.date).getDate() === new Date().getDate()
        } else if ()
     }
     `

File: scheduleCron.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+};
+ 
+async function filterInvoices(filteredSchedules) {
+    var filteredInvoices = [];
+    var invoiceFilter;
+    for(let i =0 ; i< filteredSchedules.length ; i++){
+        invoiceFilter = await invoice.find({_id : filteredSchedules[i].invoiceNumber});
```
Comment (faisal3389) [inline scheduleCron.js:77]: add the isActive condition here only

File: scheduleCron.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        try{
+            invoiceFilter.filter(function(item) {
+                if( item.isActive === true ){
+                    const newInvoice = item;
```
Comment (faisal3389) [inline scheduleCron.js:81]: const newInvoice = {...item};

File: scheduleCron.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                    newInvoice.date = new Date().toDateString();
+                    item=newInvoice;
```
Comment (faisal3389) [inline scheduleCron.js:83]: item should be immutable

File: scheduleCron.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        
+    }
+    return filteredInvoices;
+}
+const sendMail = async(filteredInvoices) => {
+    const invoices = [];
+    for (var i = 0; i < filteredInvoices.length; i++) {
+        invoices.push({
+            filename: `${filteredInvoices[i]._id}.pdf`,
+            path: __dirname.replace("routes", "Invoices") + "/" + `${filteredInvoices[i]._id}.pdf`,
```
Comment (faisal3389) [inline scheduleCron.js:102]: __dirname.InvoiceOutputs + "/" + `${filteredInvoices[i]._id}.pdf`,

Comment (faisal3389) [reply scheduleCron.js:102]: put `InvoiceOutputs` in .gitignore

File: scheduleCron.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        });
+    }
+    console.log("ready to send mails");
+    var transporter = nodemailer.createTransport({
+        service: "gmail",
+        host: "smtp.gmail.com",
+        secure: true,
+        auth: {
+            user: "codeunity.test@gmail.com",
+            pass: "beparojilxrzcwyc"
```
Comment (faisal3389) [inline scheduleCron.js:112]: use gmail api key

Comment (faisal3389) [reply scheduleCron.js:112]: use oath https://nodemailer.com/smtp/oauth2/

File: scheduleCron.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        }
+    });
+    var mailOptions = {
+        from: "CodeUnity Technologies private Limited",
+        to: "satish.chadive@gmail.com",
+        subject: "CodeUnity Technologies invoices today",
+        text: "Hello Auditor, Please find the attached invoices  below",
+        attachments: invoices,
+    };try {
+        await transporter.sendMail(mailOptions);
```
Comment (faisal3389) [inline scheduleCron.js:122]: for v1, update the status for whole array

File: scheduleCron.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        console.log(error);
+    }
+};
+const invoicePdfsGeneration = async (filteredInvoices) => {
+    for (var i = 0; i < filteredInvoices.length; i++){
+        try{
+            await generatePdf(filteredInvoices[i]);
+            var findInvoiceById=scheduleStatus.filter(obj => obj.invoiceId == filteredInvoices[i]._id);
+            findInvoiceById[0].pdfPrintStatus="true";
+            findInvoiceById[0].sendMailStatus="true";
```
Comment (faisal3389) [inline scheduleCron.js:136]: move this above

File: scheduleCron.js [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        }
+        catch(err){
+            console.log(`failed to generate pdf for invoice Id : ${filteredInvoices[i]._id}`);
+        }
+    }
+        
+};
+async function generatePdf(invoiceData) {
+    //invoice body : 
+    const invoice_data = invoiceData;
```
Comment (faisal3389) [inline scheduleCron.js:146]: fix the case

File: scheduleCron.js [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    
+    try {
+        var template_path = __dirname.replace("routes", "templates") + "/templates" + "/invoice.html";
```
Comment (faisal3389) [inline scheduleCron.js:149]: fix the replace function

File: scheduleCron.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+        var templateHtml = fs.readFileSync(template_path, "utf-8");
```
Comment (faisal3389) [inline scheduleCron.js:150]: add await to fs.readFileSync

File: scheduleCron.js [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+        var template = handlebars.compile(templateHtml);
+        var finalHtml = encodeURIComponent(template(invoice_data));
+        const pdfpath = __dirname.replace("routes", "Invoices") + "/" + `${invoiceData._id}.pdf`;
+        var options = {
```
Comment (faisal3389) [inline scheduleCron.js:154]: don't open browser in the backend, remove the below part in this function

File: scheduleCron.js [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+    } 
+    catch (err) {
+        console.error(err);
+        return false;
+    }
+}
+
+const deleteGenereatedPdfs = async (filteredInvoices) => {
+    for (var i = 0; i < filteredInvoices.length; i++) {
+        await fs.unlink(__dirname.replace("routes", "Invoices") + "/" + `${filteredInvoices[i]._id}.pdf`, function (err) {
```
Comment (faisal3389) [inline scheduleCron.js:186]: use rmDir

File: scheduleCron.js [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        });
+    }
+};
+const procedure = async() => {
+    var filteredSchedulesForToday = await filterSchedules();
+    var invoicesToSend = await filterInvoices(filteredSchedulesForToday);
+    await invoicePdfsGeneration(invoicesToSend);
+    await sendMail(invoicesToSend);
+    await deleteGenereatedPdfs(invoicesToSend);
+    ScheduleStatus.insertMany(scheduleStatus);
```
Comment (faisal3389) [inline scheduleCron.js:200]: let not use global object

File: app.js (hunk 1 above)
Comment (chadivesatish) [reply app.js:91]: Removed it , now we can run our cron job independently using the command "node scheduleCron" in the terminal

File: scheduleCron.js (hunk 2 above)
Comment (chadivesatish) [reply scheduleCron.js:2]: I have a small confusion in this, let me clarify it in the next meet and then update accordingly

**Pull request merged**